core/sql/common/csconvert.cpp

```diff
@@ -1275,3 +1279,45 @@ char * findStartOfChar( char *someByteInChar, char *startOfBuffer )
      rtnv-- ;
   return rtnv ;
 }
+
+/* A method to do character set conversion , using Glibc iconv */
+static int charsetConvert(const char *srcCharset,const char *targetCharset,char *inputbuf, size_t inputlen, char *outbuf,size_t outlen)
+{
+  char **ptrin = &inputbuf;
+  char **ptrout = &outbuf;
+
+  iconv_t cd;
+  cd = iconv_open(targetCharset,srcCharset);
+
+  if (cd==0) 
+    return -1;
+
+  if (iconv(cd,ptrin,(size_t*)&inputlen,ptrout,(size_t *)&outlen) == -1) 
+  {
+    //error occurs
+    iconv_close(cd);
+    return -1;
+  }
+
+  iconv_close(cd);
+  return outlen;
+}
+
+/* convert gbk string into UTF8 */
+int gbkToUtf8(char* gbkString, size_t gbklen, 
+              char* result ,size_t outlen, bool addNullAtEnd)
+{
+   int originalOutlen = outlen;
+   int finalLength = charsetConvert( "gbk","utf-8", gbkString, gbklen,  result, outlen);
+   
+   if (finalLength == -1 ) 
+     return -1;
+   
+   if ( addNullAtEnd )
+   {
+     if(originalOutlen >= finalLength )
+       result[finalLength] = 0;
```

Comment:
If there isn't room for a null terminator, should we return -1? (That is, does the caller always expect a null terminator to be added when "addNullAtEnd" is true?)